### ISSUE TITLE
Date value from iterator during Daylight Savings Gap Hour

### ISSUE THREAD

**Issue by toby-cisco** [mangstadt/biweekly#38]
Mike,

I would like to get your comments on a problem I am seeing regarding the `Date` returned from a `DateIterator` when the event occurs during the Daylight Savings Time (DST) "gap" hour (at least for the Pacific Time Zone) of 02:00 to 02:59. I have written a test shown below that illustrates the problem. 

Specifically, for the occurrence on the day DST begins, the Date returned has the time as 01:30 when I believe it should be 03:30. Is my understanding of how the "gap" hour is handled incorrect or is my use of the library incorrect? Or Both? Also listed below is a test that uses Java 8's ZonedDate time for the same set of dates and in it the occurrence that would fall on the "gap" hour is advanced to 03:30.

Thank you in advance for any feedback.

```java
       /**
     *  Output:
     *      Sun Mar 06 02:30:00 PST 2016
     *      Sun Mar 13 01:30:00 PST 2016 <-- Should be 03:30?
     *      Sun Mar 20 02:30:00 PDT 2016
     * @throws ParseException
     */
    @Test
    public void daylightSavingsGapHour() throws ParseException {
        final TimeZone pacificTimeZone = TimeZone.getTimeZone("America/Los_Angeles");

        // Start date: March 6, 2016 at 02:30 - one week before DST begins
        RecurrenceIterator recurrenceIterator = RecurrenceIteratorFactory.createRecurrenceIterator(
                "RRULE:FREQ=WEEKLY;INTERVAL=1;COUNT=3", new DateTimeValueImpl(2016, 3, 6, 2, 30, 0), pacificTimeZone, true);

        DateIterator dateIterator = DateIteratorFactory.createDateIterator(recurrenceIterator);
        while (dateIterator.hasNext()) {
            Date next = dateIterator.next();
            System.out.println(next);
        }
    }
```

Here is an example of date-time stamps I think should be given using Java 8's ZonedDateTime:

```java
    /**
     *  Output:
     *      2016-03-06T02:30-08:00[America/Los_Angeles]
     *      2016-03-13T03:30-07:00[America/Los_Angeles] <--- Time is at 03:30 and in DST (e.g. -07:00 offset)
     *      2016-03-20T02:30-07:00[America/Los_Angeles]
     */
    @Test
    public void gapHourUsingJavaTimePackage() {
        // Time is March 6, 2016 at 02:30 (One week before DST begins)
        final ZonedDateTime start = ZonedDateTime.of(2016, 3, 6, 2, 30, 0, 0, ZoneId.of("America/Los_Angeles"));
        System.out.println(start);

        // Add One week
        ZonedDateTime march13th2016 = start.plusWeeks(1);
        System.out.println(march13th2016);
        
        // Add two weeks
        ZonedDateTime march20th2016 = start.plusWeeks(2);
        System.out.println(march20th2016);
    }
```

**Comment by mangstadt** [comment]
Hi Toby,

Thank you for pointing this out.  A similar, possibly related, issue was reported in https://github.com/mangstadt/biweekly/issues/16.

I don't know what is causing the problem, but I believe I know a work-around.  If you specify everything in UTC time, it seems to work as expected.  This is what biweekly always does when it calls this method.  It never uses a timezone other than UTC.

```java
TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));

//March 6, 2016 at 02:30 (one week before DST begins, UTC offset -8)
DateValue startUtc = new DateTimeValueImpl(2016, 3, 6, 10, 30, 0);

String rrule = "RRULE:FREQ=WEEKLY;INTERVAL=1;COUNT=3";
TimeZone utc = TimeZone.getTimeZone("UTC");

RecurrenceIterator recurrenceIterator = RecurrenceIteratorFactory.createRecurrenceIterator(rrule, startUtc, utc, true);
DateIterator dateIterator = DateIteratorFactory.createDateIterator(recurrenceIterator);
while (dateIterator.hasNext()) {
	System.out.println(dateIterator.next());
}
```

Output:

```
Sun Mar 06 02:30:00 PST 2016
Sun Mar 13 03:30:00 PDT 2016
Sun Mar 20 03:30:00 PDT 2016
```

Is this the output you are expecting?

**Comment by toby-cisco** [comment]
Mike,

That output is not quite quite what I am looking for. The last output, for March 20th, 2016, has the time at 03:30:00. I am thinking it should return to 02:30. The use case would be this:

You have a meeting that occurs every Sunday for three Sundays and begins March 6th at 02:30:00 PST.

The occurrences would then be March 6th, March 13th, and March 20th. The first and last occurrence would be at the expected hour of 02:30:00 AM. The March 13th date would occur, I think, at 03:30:00 AM due to the normal time falling within the 'gap" hour.

Any thoughts on how to achieve this result, assuming the result I want is how things work around the "gap" hour.

**Comment by mangstadt** [comment]
Oh, I see.  So biweekly is outputting:

```
Sun Mar 06 02:30:00 PST 2016
Sun Mar 13 01:30:00 PST 2016
Sun Mar 20 02:30:00 PDT 2016
```

But it *should* be outputting

```
Sun Mar 06 02:30:00 PST 2016
Sun Mar 13 03:30:00 PST 2016
Sun Mar 20 02:30:00 PDT 2016
```

I'm not familiar with how the time within the gap hour is supposed to be handled.  Can you provide a source to where you got this information, just to confirm it is true?  Thank you.

**Comment by fge** [comment]
According to [this URL](http://www.kanzaki.com/docs/ical/rrule.html) specifying `RRULE`, the iterations should be:

```
Sun Mar 06 02:30:00 PST 2016
Sun Mar 13 02:30:00 PST 2016
Sun Mar 20 02:30:00 PDT 2016
```

`RRULE` defines _periods_ between events, not _durations_.

**Comment by toby-cisco** [comment]
@mangstadt and @fge,

I would *think* the output would be

```
Sun Mar 06 02:30:00 PST 2016
Sun Mar 13 03:30:00 PDT 2016
Sun Mar 20 02:30:00 PDT 2016
```
 The difference being the Mar 13th occurrence is in PDT (-7 offset) and not PST (-8 offset). 

@mangstadt, as far as the sources I looked at for how the "gap" hour ought to be handled, I look at three separate ones, none of which, granted, are specific to the ICal standard. First, the Java `ZonedDateTime` [documentation] (http://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html) has this to say regarding the "gap" hour:

```
For Gaps, the general strategy is that if the local date-time falls in the middle of a Gap, then the resulting zoned date-time will have a local date-time shifted forwards by the length of the Gap, resulting in a date-time in the later offset, typically "summer" time.
```

The second and third sources were MS Exchange and Google Calendar. If an event falls within the "gap" hour, those apps will show that one occurrence as happening one-hour later than what the ```dtStart``` value would dictate.

I grant that my sources are not ICalendar specific, which is what BiWeekly is implementing. But rolling a date-time forward by the gap duration does seem to be how others handle the gap.  What BiWeekly is currently doing, namely going **back** 1-hour, definitely does not seem right. 

I am trying to set aside some time today or tomorrow to dig into the code to see if I can find come up with a fix myself. If I do, I'll submit a PR for your consideration.

Thanks again.

**Comment by mangstadt** [comment]
Merged in 0470cb8867f26ca1f4861a024222bfd43529b3c3.  Thank you again.
